pulsar-broker/src/main/java/org/apache/pulsar/compaction/Compactor.java

```diff
@@ -0,0 +1,205 @@
+
+        @Override
+        public void run() {
+            RawReader reader = null;
+            try {
+                reader = RawReader.create(pulsar, topic,
+                                          COMPACTION_SUBSCRIPTION).get(10, TimeUnit.SECONDS);
+                Map<String,MessageId> latestForKey = new HashMap<>();
+
+                RawMessage firstMessage = reader.readNextAsync().get(10, TimeUnit.SECONDS);
+                MessageId firstMessageId = firstMessage.getMessageId();
+                latestForKey.put(extractKey(firstMessage), firstMessageId);
+                MessageId latestMessageId = firstMessageId;
+
+                BlockingQueue<Future<RawMessage>> futures = new ArrayBlockingQueue<Future<RawMessage>>(
+                        MAX_OUTSTANDING_READS);
+                try {
+                    while (true) { // breaks with exception when there's no more messages
+                        Future<RawMessage> f = reader.readNextAsync();
+                        if (!futures.offer(f)) {
+                            RawMessage m = futures.take().get(10, TimeUnit.SECONDS);
+                            MessageId id = m.getMessageId();
+                            latestMessageId = id;
+                            String key = extractKey(m);
+                            latestForKey.put(key, id);
+                            futures.offer(f);
+                        }
+                    }
+                } catch (TimeoutException te) {
+                    // Can't read any more
+                }
+
+                futures.forEach(Compactor::cancelOrFree);
+                futures.clear();
+
+                reader.seekAsync(firstMessageId).get(10, TimeUnit.SECONDS);
+
+                LedgerHandle lh = bk.createLedger(COMPACTED_TOPIC_ENSEMBLE_SIZE,
+                                                  COMPACTED_TOPIC_ACK_QUORUM_SIZE,
+                                                  BookKeeper.DigestType.CRC32,
+                                                  COMPACTED_TOPIC_LEDGER_PASSWORD);
+                CompletableFuture<Void> lastAddFuture = CompletableFuture.completedFuture(null);
+                while (true) { // will break when it gets message matching latestMessageId
+                    Future<RawMessage> f = reader.readNextAsync();
+                    if (!futures.offer(f)) {
+                        RawMessage m = futures.take().get(10, TimeUnit.SECONDS);
+                        MessageId id = m.getMessageId();
+                        String key = extractKey(m);
+
+                        if (latestForKey.get(key).equals(id)) {
+                            final CompletableFuture<Void> bkf = new CompletableFuture<>();
+                            ByteBuf serialized = m.serialize();
+                            lh.asyncAddEntry(serialized,
+                                             (rc, ledger, eid, ctx) -> {
+                                                 if (rc != BKException.Code.OK) {
```

Comment:
if add entry is broken in the middle of compaction, you might want to react as soon as possible. so you might consider using a field like `lastSeenException` to track this. once you see an exception, stop compaction, since the ledger is not usable again.

on seeing an exception, you have two choices here:

1) treat the comapction as failed, and delete the ledger.

2) treat the compaction also as completed, return the ledger and the last compacted message id. so next time, when you do the comapction, you don't need to restart from the begin, you can start from where you left.

1) is easy to implement, but 2) provides a better approach to tolerant failures, without scanning data again and again due to some failures.

Reply:
Implementing 1. 

2 could have problems where you end up having an error while reading through already compacted data, so you would end up losing data because the reader can't tell the difference between compacted data and non-compacted data.

Reply:
Actually, working with timeouts to find the end of the topic we'll have the same problem with as 2.